pyanaconda/ui/helpers.py

```diff
@@ -115,67 +126,79 @@ def set_source_hdd_iso(self, device_name, iso_path):
         :param iso_path: full path to the source ISO file
         :type iso_path: string
         """
-        partition = device_name
-        # the GUI source spoke also does the copy
-        old_source = copy.copy(self.data.method)
+        self._clean_hdd_iso()
 
-        # if a different partition was used previously, unprotect it
-        if old_source.method == "harddrive" and old_source.partition != partition:
-            self._clean_hdd_iso()
+        new_source_proxy = create_source(constants.SOURCE_TYPE_HDD)
+        new_source_proxy.SetPartition(device_name)
+        # the / gets stripped off by payload.ISO_image
+        new_source_proxy.SetDirectory("/" + iso_path)
 
         # protect current device_name
         mark_protected_device(device_name)
 
-        self.data.method.method = "harddrive"
-        self.data.method.partition = partition
-        # the / gets stripped off by payload.ISO_image
-        self.data.method.dir = "/" + iso_path
-
-        # as we already made the device_name protected when
-        # switching to it, we don't need to protect it here
+        payload_proxy = get_payload()
+        payload_proxy.SetSources([get_object_path(new_source_proxy)])
 
-    def set_source_url(self, url=None):
+    def set_source_url(self, url):
         """ Switch to install source specified by URL """
         # clean any old HDD ISO sources
         self._clean_hdd_iso()
```

Comment:
I would rename the method so something like `_clean_sources`. We need to call the tear-down task before we set up the sources.